### PR TITLE
Add some filters and a new Trait to consolidate the export functionality.

### PR DESCRIPTION
**Pull request by Camwyn**
The new Trait contains all the shared functionality of the two export links - which is pretty much all of it.
The old trait has been modified to not throw weird PHPCS issues about calling the class prop. 👃🏼 

There are now a slew of new filters for folks who desire to modify the links:

 - `tec_events_subscribe_link_url`
 - `tec_events_{$slug}_subscribe_link_url`
 - `tec_events_export_link_visibility`
 - `tec_events_{$slug}_export_link_visibility`
 - `tec_events_export_link_url`
 - `tec_events_{$slug}_export_link_url`
 - `tec_events_export_link_url_single`
 - `tec_events_{$slug}_export_link_url_single`

Note they all use "tec", and none use "views_v2"

[TEC-4916](https://theeventscalendar.atlassian.net/browse/TEC-4916)

PS: I apologize for the PHPCS noise, I touched a bit of older code here 😉 

[TEC-4916]: https://stellarwp.atlassian.net/browse/TEC-4916?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: readme.txt [hunk 1]
```diff
@@ -231,6 +231,11 @@ Remember to always make a backup of your database and files before updating!
 
 == Changelog ==
 
+= TBD [TBD] =
+
+* Tweak - Add new filters to allow customization of the subscribe and export links. [TEC-4916]
+* Tweak - Added filters: , `tec_events_subscribe_link_url`, `tec_events_{$slug}_subscribe_link_url`, `tec_events_export_link_visibility`, `tec_events_{$slug}_export_link_visibility`, `tec_events_export_link_url_single`, `tec_events_{$slug}_export_link_url_single`, `tec_events_export_link_url`, `tec_events_{$slug}_export_link_url`
```
Comment (stratease) [inline readme.txt:237]:
```suggestion
* Tweak - Added filters: `tec_events_subscribe_link_url`, `tec_events_{$slug}_subscribe_link_url`, `tec_events_export_link_visibility`, `tec_events_{$slug}_export_link_visibility`, `tec_events_export_link_url_single`, `tec_events_{$slug}_export_link_url_single`, `tec_events_export_link_url`, `tec_events_{$slug}_export_link_url`
```

File: src/Tribe/Views/V2/iCalendar/Traits/Outlook_Methods.php [hunk 2]
```diff
@@ -53,29 +64,69 @@ protected function generate_outlook_add_url_parameters( $calendar = 'live' ) {
...
+		 * @since TBD
+		 *
+		 * @var array    $params   The params used in the add_query_arg.
+		 * @var /WP_Post $event    The Event the link is for. As decorated by tribe_get_event().
+		 * @var string   $calendar The slug of the calendar. Values can be "outlook-365" and "outlook-live".
+		 */
+		$params = apply_filters( 'tec_events_single_event_outlook_link_parameters', $params, $event, $calendar );
 
-		$subject = $this->space_replace_and_encode( strip_tags( $event->post_title ) );
+		$service_slug = static::$service_slug;
```
Comment (stratease) [inline src/Tribe/Views/V2/iCalendar/Traits/Outlook_Methods.php:104]:
```suggestion
		$service_slug = static::get_service_slug();
```
Getter like we do elsewhere? Just a thought.

**Review by stratease — CHANGES_REQUESTED** [review]
Is there already test coverage for this? Seems like changes that would warrant an updated test.

File: src/Tribe/Views/V2/iCalendar/Traits/Outlook_Methods.php [hunk 3]
```diff
@@ -85,78 +136,87 @@ protected function generate_outlook_add_url_parameters( $calendar = 'live' ) {
...
+		 * Note: This happens before space conversion and truncation happens!
+		 *
+		 * @since TBD
+		 *
+		 * @param string $body The event description.
+		 */
+		$body = apply_filters( 'tec_events_ical_outlook_event_description_content', $body );
+
+		// Truncate the Event description and add permalink if greater than 900 characters.
+		if ( strlen( $body ) > 900 ) {
```
Comment (stratease) [inline src/Tribe/Views/V2/iCalendar/Traits/Outlook_Methods.php:159]:
```suggestion
		if ( strlen( $body ) > $filtered_body_length ) {
```
Shouldn't we filter this so people can override? Seems like the type of thing we shouldn't hard code.

**Review by stratease — CHANGES_REQUESTED** [review]
Some thoughts and potential tweaks.

File: src/Tribe/Views/V2/iCalendar/Traits/Export_Link.php [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	 * Getter function for the uri property.
+	 *
+	 * @since 5.12.0
+	 *
+	 * @param View|null $view The current View object.
+	 *
+	 * @return string The url for the link calendar subscription "feed", or download.
+	 */
+	public function get_uri( View $view = null ) {
+		$is_single = null === $view || is_single( Tribe__Events__Main::POSTTYPE );
```
[the remark this answers is not in the document]
Comment (Camwyn) [reply src/Tribe/Views/V2/iCalendar/Traits/Export_Link.php:97]: A few of the classes (which override this) already ahve the same logic. I don't disagree but I think that's a fix for another day.

File: src/Tribe/Views/V2/iCalendar/Traits/Export_Link.php [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+		 * @param boolean       $visible     Whether to display the link.
+		 * @param Link_Abstract $link_object The current link object.
+		 */
+		return apply_filters( "tec_events_{$slug}_export_link_visibility", $visible, $this );
+	}
+
+	/**
+	 * Getter function for the uri property.
+	 *
+	 * @since 5.12.0
```
[the remark this answers is not in the document]
Comment (Camwyn) [reply src/Tribe/Views/V2/iCalendar/Traits/Export_Link.php:90]: Yeah, I was moving/duplicating some - this is sufficiently new I think it should be TBD.

File: src/Tribe/Views/V2/iCalendar/Traits/Outlook_Methods.php (hunk 3 above)
Comment (Camwyn) [reply src/Tribe/Views/V2/iCalendar/Traits/Outlook_Methods.php:159]: This is an old value and my understanding is it's a limit of the API - perhaps we should revisit it (later)?

**Pull request closed without merging**